### PR TITLE
Fix missing error messages in Managed SNI

### PR DESCRIPTION
**Pull request by cheenamalhotra**
Found these error messages not included in exceptions while working on Managed SNI.
With these changes, empty messages will not be thrown and error codes shall correspond to message text.

### CODE REVIEW

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNINpHandle.cs [hunk 1]
```diff
@@ -38,15 +37,14 @@ internal sealed class SNINpHandle : SNIPhysicalHandle
...
-        public SNINpHandle(string serverName, string pipeName, long timerExpire, object callbackObject)
+        public SNINpHandle(string serverName, string pipeName, long timerExpire)
         {
             long scopeID = SqlClientEventSource.Log.TrySNIScopeEnterEvent("<sc.SNI.SNINpHandle.SNINpHandle |SNI|INFO|SCOPE> Constructor");
             SqlClientEventSource.Log.TrySNITraceEvent("<sc.SNI.SNINpHandle.SNINpHandle |SNI|INFO> Constructor. server name = {0}, pipe name = {1}", serverName, pipeName);
             try
             {
                 _sendSync = new object();
                 _targetServer = serverName;
-                _callbackObject = callbackObject;
```
Comment (karinazhou) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNINpHandle.cs:49]: Why is this removed?

Comment (cheenamalhotra) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNINpHandle.cs:49]: It's not needed, never used.

File: src/Microsoft.Data.SqlClient/netcore/src/Resources/Strings.resx [hunk 2]
```diff
@@ -931,7 +931,7 @@
     <value>Associating port with I/O completion mechanism failed</value>
   </data>
   <data name="SNI_ERROR_11" xml:space="preserve">
-    <value>Timeout error</value>
+    <value>Connection timed out.</value>
```
Comment (karinazhou) [inline src/Microsoft.Data.SqlClient/netcore/src/Resources/Strings.resx:934]: You may change this in src\Microsoft.Data.SqlClient\netfx\src\Resources\Strings.resx to keep consistency.

Comment (cheenamalhotra) [reply src/Microsoft.Data.SqlClient/netcore/src/Resources/Strings.resx:934]: I didn't want to change NetFx though.. 🤔 
Might as well revert it for now.

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNINpHandle.cs [hunk 3]
```diff
@@ -86,7 +84,7 @@ public SNINpHandle(string serverName, string pipeName, long timerExpire, object
 
                 if (!_pipeStream.IsConnected || !_pipeStream.CanWrite || !_pipeStream.CanRead)
                 {
-                    SNICommon.ReportSNIError(SNIProviders.NP_PROV, 0, SNICommon.ConnOpenFailedError, string.Empty);
+                    SNICommon.ReportSNIError(SNIProviders.NP_PROV, 0, SNICommon.ConnOpenFailedError, Strings.SNI_ERROR_1);
```
Comment (karinazhou) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNINpHandle.cs:87]:
```suggestion
                    SNICommon.ReportSNIError(SNIProviders.NP_PROV, 0, SNICommon.ConnOpenFailedError, Strings.SNI_ERROR_40);
```

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNINpHandle.cs (hunk 1 above)
Comment (karinazhou) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNINpHandle.cs:49]: If that is the case, should we also remove it from its caller:
private SNINpHandle CreateNpHandle(DataSource details, long timerExpire, object callbackObject, bool parallel) ?

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNINpHandle.cs [hunk 4]
```diff
@@ -38,15 +37,14 @@ internal sealed class SNINpHandle : SNIPhysicalHandle
         private int _bufferSize = TdsEnums.DEFAULT_LOGIN_PACKET_SIZE;
         private readonly Guid _connectionId = Guid.NewGuid();
 
-        public SNINpHandle(string serverName, string pipeName, long timerExpire, object callbackObject)
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNINpHandle.cs:41]: The `callbackObject` parameter can be removed in the following locations:

- `SNITcpHandle` ctor
- `SNIProxy.CreateNpHandle()`
- `SNIProxy.CreateTcpHandle()`
- `SNIProxy.CreateConnectionHandle()`

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNINpHandle.cs (hunk 1 above)
Comment (cheenamalhotra) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNINpHandle.cs:49]: Done as per Davoud's above comment.

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNINpHandle.cs [hunk 5]
```diff
@@ -126,7 +124,7 @@ public override int ProtocolVersion
             {
                 try
                 {
-                    return (int)_sslStream.SslProtocol;
+                    return (int)_sslStream?.SslProtocol;
```
[the remark this answers is not in the document]
Comment (cheenamalhotra) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNINpHandle.cs:127]: stream gets null sometimes, that's why I added this here. I did see the catch but it's very generic - are we sure we want to let Null object reference errors handled by base?

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/LocalDB.Windows.cs [hunk 6]
```diff
@@ -88,6 +88,25 @@ internal static uint MapLocalDBErrorStateToCode(LocalDBErrorState errorState)
...
+                case LocalDBErrorState.INVALID_CONFIG:
+                    return Strings.SNI_ERROR_53;
+                case LocalDBErrorState.NO_SQLUSERINSTANCEDLL_PATH:
+                    return Strings.SNI_ERROR_54;
+                case LocalDBErrorState.INVALID_SQLUSERINSTANCEDLL_PATH:
+                    return Strings.SNI_ERROR_55;
+                case LocalDBErrorState.NONE:
+                    return Strings.SNI_ERROR_50;
+                default:
+                    return Strings.SNI_ERROR_53;
```
[the remark this answers is not in the document]
Comment (cheenamalhotra) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/LocalDB.Windows.cs:106]: At this particular place it's not needed, there are too many logs where this gets called.

[the remark this answers is not in the document]
Comment (cheenamalhotra) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/LocalDB.Windows.cs:106]: Also for traces, I'm already working on all changes, this PR was mainly to fix error messages and minor cleanups that make up a lot to review otherwise.

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNINpHandle.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (DavoudEshtehari) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNINpHandle.cs:127]: The `zero` interprets to `SslProtocols.None` that ignores in generating the warning message at **TdsParser**. For this purpose, it works unless you've been considering something more for its usage that needs more changes in both **Net Core** and **Net Fx**.

[the remark this answers is not in the document]
Comment (cheenamalhotra) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNINpHandle.cs:127]: Ok

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/LocalDB.Windows.cs [hunk 7]
```diff
@@ -117,15 +136,15 @@ private bool LoadUserInstanceDll()
                     // If there was no DLL path found, then there is an error.
                     if (dllPath == null)
                     {
-                        SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.INVALID_PROV, 0, MapLocalDBErrorStateToCode(registryQueryErrorState), string.Empty);
+                        SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.INVALID_PROV, 0, MapLocalDBErrorStateToCode(registryQueryErrorState), MapLocalDBErrorStateToErrorMessage(registryQueryErrorState));
                         SqlClientEventSource.Log.TrySNITraceEvent("<sc.SNI.LocalDB.Windows.LoadUserInstanceDll |SNI|ERR >User instance DLL path is null.");
```
Comment (JRahnama) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/LocalDB.Windows.cs:140]: Since we are improving the logs and tracings, isn't it better to change message to what you did in [PR#867](https://github.com/dotnet/SqlClient/pull/867)?

**Review by karinazhou — APPROVED** [review]

**Review by DavoudEshtehari — APPROVED** [review]

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/LocalDB.Windows.cs [hunk 8]
```diff
@@ -88,6 +88,25 @@ internal static uint MapLocalDBErrorStateToCode(LocalDBErrorState errorState)
             }
         }
 
+        internal static string MapLocalDBErrorStateToErrorMessage(LocalDBErrorState errorState)
```
Comment (johnnypham) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/LocalDB.Windows.cs:91]: Could use a switch expression here?

Comment (JRahnama) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/LocalDB.Windows.cs:91]: Except the minor comment it LGTM.

**Review by JRahnama — APPROVED** [review]

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/LocalDB.Windows.cs (hunk 7 above)
Comment (cheenamalhotra) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/LocalDB.Windows.cs:140]: To be done in Traces PR, this PR is not targeted to change any traces.
